### PR TITLE
coll: TSP gatherv linear

### PR DESCRIPTION
**Pull request by tarudoodi**
Implementation and tests for linear gatherv. 
Includes gentran implementation for issend.

### CODE REVIEW

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: test/mpi/coll/test_coll_algos.sh [hunk 1]
```diff
@@ -262,4 +262,23 @@ for algo_name in ${algo_names}; do
...
+testing_env+="env=MPIR_CVAR_IGATHERV_DEVICE_COLLECTIVE=0 "
+algo_names="gentran_linear"
+
+for algo_name in ${algo_names}; do
+    for kval in ${kvalues}; do
+        #set the environment
+        env="${testing_env} env=MPIR_CVAR_IGATHERV_INTRA_ALGORITHM=${algo_name} "
+
+        coll_algo_tests+="gatherv 5 ${env}${nl}"
+    done
```
Comment (hzhou) [inline test/mpi/coll/test_coll_algos.sh]: I could not find the test `gatherv.c`, where was this added?

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/coll/transports/gentran/gentran_utils.c [hunk 2]
```diff
@@ -105,6 +105,25 @@ static void vtx_issue(int vtxid, MPII_Genutil_vtx_t * vtxp, MPII_Genutil_sched_t
                 }
                 break;
 
+            case MPII_GENUTIL_VTX_KIND__ISSEND:{
+                    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+
+                    MPIC_Isend(vtxp->u.issend.buf,
```
Comment (hzhou) [inline src/mpi/coll/transports/gentran/gentran_utils.c]: Should this be `MPIC_Issend`?

File: test/mpi/coll/test_coll_algos.sh (hunk 1 above)
Comment (tarudoodi) [reply test/mpi/coll/test_coll_algos.sh]: You're right there isn't any test called `gatherv.c` and AFAIK it wasn't added either. I will update the test here. I didn't check it when I created the PR for Akhil's commits.

File: src/mpi/coll/transports/gentran/gentran_utils.c (hunk 2 above)
Comment (tarudoodi) [reply src/mpi/coll/transports/gentran/gentran_utils.c]: Yes.

File: test/mpi/coll/test_coll_algos.sh (hunk 1 above)
Comment (tarudoodi) [reply test/mpi/coll/test_coll_algos.sh]: So there was a commit that renamed `coll3` test to `gatherv` and I missed that. I added that commit now.

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
ch4:ucx failed 1 test with `embedded` configure for the `ialltoallv scattered` test: 

    Error Message
    not ok 477 - ./coll/alltoallv0 10

    Directory: ./coll
    File: alltoallv0
    Num-procs: 10
    Timeout: 360
    Date: "Wed Feb 27 10:52:12 2019"
    ...
    ## Test output (expected 'No Errors'):
    ## [bb80:23298:0:23298] ucp_proxy_ep.c:215  UCX Assertion `proxy_ep->uct_ep != ((void *)0)' 

Last nightly didn't catch this. Need investigate.

**Comment by hzhou** [comment]
Re-run the test to confirm:
test:jenkins/ch4/ucx

**Comment by tarudoodi** [comment]
> ch4:ucx failed 1 test with `embedded` configure for the `ialltoallv scattered` test:
> 
> ```
> Error Message
> not ok 477 - ./coll/alltoallv0 10
> 
> Directory: ./coll
> File: alltoallv0
> Num-procs: 10
> Timeout: 360
> Date: "Wed Feb 27 10:52:12 2019"
> ...
> ## Test output (expected 'No Errors'):
> ## [bb80:23298:0:23298] ucp_proxy_ep.c:215  UCX Assertion `proxy_ep->uct_ep != ((void *)0)' 
> ```
> Last nightly didn't catch this. Need investigate.

@hzhou,  should we track this in the separate issue #3616  that you created?  This PR now passes all the tests.

**Comment by hzhou** [comment]
@tarudoodi  I requested some comments change. You don't need to re-run tests after that.

File: src/mpi/coll/igatherv/igatherv_tsp_linear_algos.h [hunk 3]
```diff
@@ -0,0 +1,141 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
```
[the remark this answers is not in the document]
Comment (tarudoodi) [reply src/mpi/coll/igatherv/igatherv_tsp_linear_algos.h]: @hzhou I think we don't change the year here, I tried looking for a recent example but didn't find it. I will update it on line# 7 below. Is that ok?

[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/coll/igatherv/igatherv_tsp_linear_algos.h]: You created this file in 2018 or 2019. It is new so what do you mean by "change"? The file doesn't exist and the code doesn't exist in 2006, so it is simply lie to claim a copyright in 2006.

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
@tarudoodi I see you updated the code. Are the changes regarding comments that I made?

File: src/mpi/coll/igatherv/igatherv_tsp_linear_algos.h [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    int comm_size, rank;
+    MPI_Aint extent;
+    int min_procs;
+    int tag;
+
+    rank = comm_ptr->rank;
+
+    /* For correctness, transport based collectives need to get the
+     * tag from the same pool as schedule based collectives */
+    mpi_errno = MPIR_Sched_next_tag(comm_ptr, &tag);
```
[the remark this answers is not in the document]
Comment (tarudoodi) [reply src/mpi/coll/igatherv/igatherv_tsp_linear_algos.h]: This comment isn't actually needed, there isn't anything special about the tags, the same function is used to get tags at other places.

**Comment by tarudoodi** [comment]
> @tarudoodi I see you updated the code. Are the changes regarding comments that I made?

Yes, it addresses those comments.

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by tarudoodi** [comment]
@hzhou  Have you seen similar build failures with `ch4/ucx  strict` build before? This seems to be happening on few other PRs as well.

**Comment by hzhou** [comment]
@tarudoodi You can ignore that for now. I believe someone is working on fixing it.

**Comment by hzhou** [comment]
@tarudoodi Also, you don't need to re-run the tests unless you changed the code. I will review the comments and documentation changes and approve the changes.  I need get a couple refactor PRs reviewed and merged in before committing additional algorithm PRs. We will rerun the tests right before merging time.

**Comment by tarudoodi** [comment]
@hzhou There was a change where I corrected the `MPIC_Isend` to `MPIC_Issend`. And the commit was added where the test was renamed. Other than that there weren't any changes.

**Comment by tarudoodi** [comment]
@hzhou  I rebased this on top of master. We can test this once and then can we merge and close it?

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by tarudoodi** [comment]
I just rebased this.

**Comment by hzhou** [comment]
test:jenkins/ch4/most

**Comment by hzhou** [comment]
@tarudoodi I think this is ready to merge. Could you remove the test_coll_gen.sh commit (and add it to the testlist pr)?

**Comment by tarudoodi** [comment]
Rebased and removed the test commit.

**Pull request merged**